### PR TITLE
curl_command_generator created

### PR DESCRIPTION
**Pull request by haseebeqx**
as previous pull request contains an unwanted commit, i created another pull request

### CODE REVIEW

**Comment by haseebeqx** [comment]
function addSlashes() actually  converts  anything in body to valid string. as it is needed for terminal and curl. server will get exact body ayway

**Comment by thc202** [comment]
I know what is doing and why is being used although, the server might not get the exact body (try sending a request with a null character, like AMF messages have).

**Comment by haseebeqx** [comment]
It's for curl .
null byte or any ASCII code gets converted to string form . like \n \r
,\u00 etc. You can test it.

**Comment by haseebeqx** [comment]
That's why it uses json stringify to add slashes instead of regular
expression
And a --data-binary to pass data. anyway server will get exactly what we
want to send

**Comment by thc202** [comment]
Did you try sending a request with null character? Maybe I'm getting different results.

**Comment by haseebeqx** [comment]
I get \u0000 for null byte. What type of request you are sending

**Comment by thc202** [comment]
I'm getting \u0000 as well but that's not being sent to the server. I tested with AMF messages but can be reproduced with a "plain" request with just a null char.
If I send the request with ZAP, the server sees the null char, if I send it with the output of the script the server sees nothing (guess it's a command line issue).

**Comment by haseebeqx** [comment]
I don't know why that happened. I did my best.

File: targeted/curl_command_generator.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+//it will generate and copy curl command based on the request
+//released under the Apache v2.0 licence.
+//You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+//author:@haseebeqx
+
+function invokeWith(msg) {
+	string = "curl -i -s -k -X  '"+msg.getRequestHeader().getMethod()+"'  \\\n";
+	header = msg.getRequestHeader().getHeadersAsString();
+	header = header.split(msg.getRequestHeader().getLineDelimiter());
+	for(i=1;i<header.length;i++){
```
Comment (thc202) [inline targeted/curl_command_generator.js]: The ```i``` needs to be set to 0 (since it's obtaining just the headers now).

**Comment by thc202** [comment]
That's probably an issue with command line and it's just a good thing to have, so, no worries.

**Comment by haseebeqx** [comment]
are you on mac. worked with kali

**Comment by thc202** [comment]
I see the same behaviour in Kali:
```
# curl --version
curl 7.26.0 (x86_64-pc-linux-gnu) libcurl/7.26.0 OpenSSL/1.0.1e zlib/1.2.7 libidn/1.25 libssh2/1.4.2 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp 
Features: Debug GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP 
```
Sending a null char:
```
# curl -i -s -k -X  'POST' -H 'Content-Type: application/x-amf'  -H 'Content-Length: 1'  --data-binary $'\u0000' 'http://localhost/'  --trace -
== Info: About to connect() to localhost port 80 (#0)
== Info:   Trying ::1...
== Info: connected
== Info: Connected to localhost (::1) port 80 (#0)
=> Send header, 126 bytes (0x7e)
0000: 50 4f 53 54 20 2f 20 48 54 54 50 2f 31 2e 31 0d POST / HTTP/1.1.
0010: 0a 55 73 65 72 2d 41 67 65 6e 74 3a 20 63 75 72 .User-Agent: cur
0020: 6c 2f 37 2e 32 36 2e 30 0d 0a 48 6f 73 74 3a 20 l/7.26.0..Host: 
0030: 6c 6f 63 61 6c 68 6f 73 74 0d 0a 41 63 63 65 70 localhost..Accep
0040: 74 3a 20 2a 2f 2a 0d 0a 43 6f 6e 74 65 6e 74 2d t: */*..Content-
0050: 54 79 70 65 3a 20 61 70 70 6c 69 63 61 74 69 6f Type: applicatio
0060: 6e 2f 78 2d 61 6d 66 0d 0a 43 6f 6e 74 65 6e 74 n/x-amf..Content
0070: 2d 4c 65 6e 67 74 68 3a 20 31 0d 0a 0d 0a       -Length: 1....
== Info: additional stuff not fine transfer.c:1037: 0 0
[...]
== Info: additional stuff not fine transfer.c:1037: 0 0
== Info: HTTP 1.1 or later with persistent connection, pipelining supported
<= Recv header, 26 bytes (0x1a)
0000: 48 54 54 50 2f 31 2e 31 20 34 30 30 20 42 61 64 HTTP/1.1 400 Bad
0010: 20 52 65 71 75 65 73 74 0d 0a                    Request..
HTTP/1.1 400 Bad Reques
```

Sending other char works fine:
```
# curl -i -s -k -X  'POST' -H 'Content-Type: application/x-amf'  -H 'Content-Length: 1'  --data-binary $'\u0001' 'http://localhost' --trace -
== Info: About to connect() to localhost port 80 (#0)
== Info:   Trying ::1...
== Info: connected
== Info: Connected to localhost (::1) port 80 (#0)
=> Send header, 126 bytes (0x7e)
0000: 50 4f 53 54 20 2f 20 48 54 54 50 2f 31 2e 31 0d POST / HTTP/1.1.
0010: 0a 55 73 65 72 2d 41 67 65 6e 74 3a 20 63 75 72 .User-Agent: cur
0020: 6c 2f 37 2e 32 36 2e 30 0d 0a 48 6f 73 74 3a 20 l/7.26.0..Host: 
0030: 6c 6f 63 61 6c 68 6f 73 74 0d 0a 41 63 63 65 70 localhost..Accep
0040: 74 3a 20 2a 2f 2a 0d 0a 43 6f 6e 74 65 6e 74 2d t: */*..Content-
0050: 54 79 70 65 3a 20 61 70 70 6c 69 63 61 74 69 6f Type: applicatio
0060: 6e 2f 78 2d 61 6d 66 0d 0a 43 6f 6e 74 65 6e 74 n/x-amf..Content
0070: 2d 4c 65 6e 67 74 68 3a 20 31 0d 0a 0d 0a       -Length: 1....
=> Send data, 1 bytes (0x1)
0000: 01                                              .
== Info: upload completely sent off: 1 out of 1 bytes
== Info: additional stuff not fine transfer.c:1037: 0 0
== Info: HTTP 1.1 or later with persistent connection, pipelining supported
<= Recv header, 17 bytes (0x11)
0000: 48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0d HTTP/1.1 200 OK.
0010: 0a                                              .
HTTP/1.1 200 OK
```

We could worry about that after merging. Could you fixup the last commit?

**Comment by haseebeqx** [comment]
I don't understand what that means

**Comment by haseebeqx** [comment]
I don't understand what you said about the last commit. I am currently at
work place. Only basic phone is available until next month

**Comment by thc202** [comment]
Fold the commits into one:
https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode

**Comment by haseebeqx** [comment]
Do you mean different scripts into one commit

**Comment by haseebeqx** [comment]
I don't have  more  commits. I use github website's interface for adding a
pull request. So last commit only counts. (I think so)

**Comment by thc202** [comment]
There are currently two commits in the PR. You can fold them with rebase interactive in your local repo, after that you just need to push and PR should be updated with just that commit.

**Pull request closed without merging**